### PR TITLE
Switch test-infra to use go modules

### PR DESCRIPTION
**Pull request by fejta**
Similar to https://github.com/kubernetes/org/pull/359
/hold

* Create `//:go` target, as well as `//:update-minor` and `//:update-patch` targets for updating deps.
* Get `kazel`, `misspell` and `golint` from `@io_kubernetes_build`, `@com_github_client9_misspell` and `@com_github_golang_lint` repos
* Run `go mod init` to create `k8s.io/test-infra` module
* Make `hack/update-deps.sh` use `go mod` instead of `dep`
  - Runs `go mod tidy && go mod vendor` to manage vendor dir
  - Runs `//:gazelle -- update-repos --from-file=go.mod` to manage `go_repositories()`
  - Manually prunes vendor dir
  - Now requires bazel to run (but ensures we're using the right golang version)
* Make `hack/update-codegen.sh` run generators manually
  - `go get` to install any missing generators
  - No longer depend on a pruned `vendor/k8s.io/code-generator/generate-groups.sh` script

Required three fixes to build and test `//... -//vendor/...`:
* The misspell fix above (nothing imports this, so `go mod tidy` deletes it
* logrus no longer prints a space before ending the line (brittle `//prow/entrypoint/run_test.go`)
* `OAuthClient.Exchange()` now requires an `...AuthCodeOption` argument

/assign @cjwagner @stevekuznetsov @spiffxp @BenTheElder

### CODE REVIEW

File: hack/update-bazel.sh [hunk 1]
```diff
@@ -17,28 +17,10 @@ set -o errexit
...
 mkdir -p ./vendor
-touch "./vendor/BUILD.bazel"
-
-if ! which bazel &> /dev/null; then
-  echo "Bazel is the preferred way to build and test the test-infra repo." >&2
-  echo "Please install bazel at https://bazel.build/ (future commits may require it)" >&2
-  deprecated-update
-  exit 0
-fi
+find vendor -name BUILD -exec rm '{}' +
```
Comment (ixdy) [inline hack/update-bazel.sh]: I'm guessing `go mod tiny` doesn't prune these?

Comment (ixdy) [reply hack/update-bazel.sh]: also minor nit: could combine these with `find vendor -name BUILD -o -name BUILD.bazel -exec '{}' +`

Comment (ibrasho) [reply hack/update-bazel.sh]: `go mod tidy` is quite different from `dep prune`. It ensures that `go.mod` and `go.sum` are up-to-date with modules on disk. It mainly adds missing moduels, and remove unused ones.

**Comment by ibrasho** [comment]
Same comment from the PR to k/test-infra:
Can't we remove the vendor/ directory from the repository now?

**Comment by ibrasho** [comment]
Also, just in case the repo is within GOPATH, `GO111MODULE=on` should be added somewhere to force using modules.

File: hack/update-deps.sh [hunk 2]
```diff
@@ -26,24 +26,12 @@ set -o errexit
...
-# See https://github.com/kubernetes/steering/issues/57
-rm -rf vendor/github.com/docker/docker/contrib
-bazel run //:dep -- ensure -v "$@"
-rm -rf vendor/github.com/golang/dep/internal/fs/testdata
-rm -rf vendor/github.com/jteeuwen/go-bindata/testdata
-rm -rf vendor/github.com/docker/docker/contrib
+
+rm -rf vendor
+bazel run //:go -- mod tidy
+bazel run //:go -- mod vendor
```
Comment (ibrasho) [inline hack/update-deps.sh]: Can we make these:
```bash
GO111MODULE=on bazel run //:go -- mod tidy
GO111MODULE=on bazel run //:go -- mod vendor
```

or export `GO111MODULE=on` earlier in this script.

Comment (fejta) [reply hack/update-deps.sh]: Done

**Comment by fejta** [comment]
> Can't we remove the vendor/ directory from the repository now?

I think we should do that later

File: hack/update-bazel.sh (hunk 1 above)
Comment (fejta) [reply hack/update-bazel.sh]: Done

File: hack/update-deps.sh [hunk 3]
```diff
@@ -26,24 +26,12 @@ set -o errexit
...
-# makes bazel sad: https://github.com/golang/dep/pull/1412
-# dep should probably be removing it, but it doesn't:
-# https://github.com/golang/dep/issues/1580
-rm -rf vendor/github.com/golang/dep/internal/fs/testdata
-# go-bindata does too, and is not maintained ...
-rm -rf vendor/github.com/jteeuwen/go-bindata/testdata
-# docker has a contrib dir with nothing we use in it, dep will retain the licenses
-# which includes some GPL, so we manually prune this. 
-# See https://github.com/kubernetes/steering/issues/57
-rm -rf vendor/github.com/docker/docker/contrib
```
[the remark this answers is not in the document]
Comment (fejta) [reply hack/update-deps.sh]: Yes, this dir doesn't seem to exist

[the remark this answers is not in the document]
Comment (BenTheElder) [reply hack/update-deps.sh]: ok thanks, `dep`'s license preserving behavior would leave behind this non-golang dir with a license file and the license audit was unhappy with that :man_shrugging:

**Comment by BenTheElder** [comment]
> Also, just in case the repo is within GOPATH, GO111MODULE=on should be added somewhere to force using modules.

please do not do that! IIRC this will almost definitely break kubetest + hack/e2e.go in k/k.

**Comment by BenTheElder** [comment]
let's make this a smooth and gradual transition and not break all the things :grimacing: forcing that mode on outside of the update tools or deleting vendor is a bit much for the first PR.

File: go.mod [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	github.com/fsouza/fake-gcs-server v0.0.0-20180612165233-e85be23bdaa8
+	github.com/ghodss/yaml v1.0.0
+	github.com/go-ini/ini v1.32.0
+	github.com/go-openapi/jsonpointer v0.0.0-20170102174223-779f45308c19
+	github.com/go-openapi/jsonreference v0.0.0-20161105162150-36d33bfe519e
+	github.com/go-openapi/spec v0.0.0-20171219195406-fa03337d7da5
+	github.com/go-openapi/swag v0.0.0-20171111214437-cf0bdb963811
+	github.com/go-sql-driver/mysql v0.0.0-20160411075031-7ebe0a500653
+	github.com/gogo/protobuf v0.0.0-20171007142547-342cbe0a0415
+	github.com/golang/dep v0.5.0
```
Comment (BenTheElder) [inline go.mod]: do we still need dep in the dependencies?

Comment (BenTheElder) [reply go.mod]: nevermind, removed in a later commit :+1:

**Comment by fejta** [comment]
Anyone wanna LGTM?

**Comment by ixdy** [comment]
can you update the docs (`docs/dep.md`) as well?

**Comment by fejta** [comment]
> update the docs (docs/dep.md) 
Done

**Comment by fejta** [comment]
Oh! A real error, not a flake. Should be fixed now
/hold cancel

File: hack/update-codegen.sh [hunk 5]
```diff
@@ -17,8 +17,69 @@ set -o errexit
...
-  prowjobs:v1
+cd "$(git rev-parse --show-toplevel)"
+
+export GOPATH="${GOPATH:-$HOME/go}"
+export PATH="$GOPATH/bin:$PATH"
+export GO111MODULE=on
+
+ensure-in-gopath() {
+  if [[ "$PWD" != "$GOPATH/src/k8s.io/test-infra" ]]; then
+    echo Sadly, $(basename "$0") most run inside GOPATH=$GOPATH, not $PWD >&2
```
Comment (ixdy) [inline hack/update-codegen.sh]: nit: most -> must

File: hack/update-codegen.sh [hunk 6]
```diff
@@ -17,8 +17,69 @@ set -o errexit
...
+    exit 1
+  fi
+}
+
+codegen-init() {
+  echo "Ensuring generators exist..." >&2
+  local ver=b1289fc74931d4b6b04bd1a259acfc88a2cb0a66
```
Comment (ixdy) [inline hack/update-codegen.sh]: where does this version come from? should it be more visible?

File: hack/update-codegen.sh [hunk 7]
```diff
@@ -17,8 +17,69 @@ set -o errexit
...
+  which deepcopy-gen &>/dev/null || go get k8s.io/code-generator/cmd/deepcopy-gen@$ver
```
Comment (ixdy) [inline hack/update-codegen.sh]: this makes no guarantees that the right version of each of these is installed in $PATH.

File: hack/update-codegen.sh [hunk 8]
```diff
@@ -17,8 +17,69 @@ set -o errexit
...
+  which client-gen &>/dev/null || go get k8s.io/code-generator/cmd/client-gen@$ver
+  which lister-gen &>/dev/null || go get k8s.io/code-generator/cmd/lister-gen@$ver
+  which informer-gen &>/dev/null || go get k8s.io/code-generator/cmd/informer-gen@$ver
+  bazel run //:go -- mod tidy
+}
+
+gen-deepcopy() {
```
Comment (ixdy) [inline hack/update-codegen.sh]: WTB: bazelified versions of these

File: hack/update-codegen.sh [hunk 9]
```diff
@@ -17,8 +17,69 @@ set -o errexit
...
+gen-lister() {
+  echo "Generating lister..." >&2
+  lister-gen \
+    --go-header-file hack/boilerplate/boilerplate.generated.go.txt \
+    --input-dirs k8s.io/test-infra/prow/apis/prowjobs/v1 \
+    --output-package k8s.io/test-infra/prow/client/listers
+}
+
+gen-informer() {
+  echo "Generating infromer..." >&2
```
Comment (ixdy) [inline hack/update-codegen.sh]: informer

File: hack/update-deps.sh [hunk 10]
```diff
@@ -26,24 +26,28 @@ set -o errexit
...
+  find vendor -type f \
+    -not -iname "*.go" \
+    -not -iname "*.proto" \
+    -not -iname "*.s" \
+    -not -iname "*.h" \
+    -not -iname "*.c" \
+    -not -iname "LICENSE*" \
+    -not -iname "AUTHORS*" \
+    -not -iname "CONTRIBUTORS*" \
+    -not -iname "LICENSE*" \
```
Comment (ixdy) [inline hack/update-deps.sh]: this one is duplicated?

also we should probably keep `COPYING` files.

Comment (ixdy) [reply hack/update-deps.sh]: and `NOTICE` files it seems

Comment (ibrasho) [reply hack/update-deps.sh]: If we want to replicate `dep`'s pruning logic, here are the regexes for files kept after pruning:

* `license*`
* `licence*`
* `copying*`
* `unlicense*`
* `copyright*`
* `copyleft*`
* `*authors*`
* `*contributors*`
* `*legal*`
* `*notice*`
* `*disclaimer*`
* `*patent*`
* `*third-party*`
* `*thirdparty*`

Comment (ibrasho) [reply hack/update-deps.sh]: If we can convert this to a Go script 😎 , we can import `gps` from `dep` and still apply the same logic (especially deleteing non-Go files).

File: hack/update-codegen.sh (hunk 5 above)
Comment (fejta) [reply hack/update-codegen.sh]: Done

File: hack/update-codegen.sh (hunk 6 above)
Comment (fejta) [reply hack/update-codegen.sh]: The version we're using now. No

File: hack/update-codegen.sh (hunk 7 above)
Comment (fejta) [reply hack/update-codegen.sh]: Correct

File: hack/update-codegen.sh (hunk 9 above)
Comment (fejta) [reply hack/update-codegen.sh]: Done

File: hack/update-deps.sh (hunk 10 above)
Comment (fejta) [reply hack/update-deps.sh]: Done

As for making the pruning perfect (as opposed to functional) let's save that for future PRs and/or creating a reusable version that lives in kubernetes/repo-infra

File: hack/update-codegen.sh (hunk 7 above)
Comment (BenTheElder) [reply hack/update-codegen.sh]: hrm, I suspect this might bite later?

Comment (fejta) [reply hack/update-codegen.sh]: True, and easy enough to fix later with a new commit? Aka we could just delete the `which ||` prefix and always run it, we could put it into bazel, etc. 

Think it is fine to deal with that later

**Comment by BenTheElder** [comment]
I think we can leverage the tricks here https://github.com/golang/go/issues/25922 to sort out tool dependencies further

**Comment by stevekuznetsov** [comment]
Go changes here LGTM

**Pull request merged**